lib/src/component_declaration/function_component.dart

```diff
@@ -0,0 +1,181 @@
+// Copyright 2020 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ignore_for_file: prefer_generic_function_type_aliases
+library over_react.component_declaration.function_component;
+
+import 'package:js/js_util.dart';
+import 'package:meta/meta.dart';
+import 'package:over_react/over_react.dart';
+import 'package:react/react.dart' as react;
+import 'package:react/react_client.dart';
+import 'package:react/react_client/js_backed_map.dart';
+
+export 'component_type_checking.dart'
+    show isComponentOfType, isValidElementOfType;
+
+/// A React component declared using a function that takes in [props] and returns rendered output.
+///
+/// See <https://facebook.github.io/react/docs/components-and-props.html#functional-and-class-components>.
+typedef dynamic UiFunctionComponent<T extends UiProps>(T props);
+
+/// Returns a `UiFactory<TProps>` for [functionComponent].
+///
+/// __Example__:
+/// ```dart
+/// UiFactory<FooProps> Foo = uiFunction((props) {
+///     // Set default props using null-aware operators.
+///     final isDisabled = props.isDisabled ?? false;
+///     final items = props.items ?? [];
+///
+///     // Return the rendered component contents here.
+///     return Dom.div()(items);
+///   },
+///   // The generated props config will match the factory name.
+///   $FooPropsConfig,
```

Comment:
We should add `displayName` if that is the only way to make it so that it shows up as `Foo` in the dev tools.

Reply:
It's not; `$FooPropsConfig` contains that information